File: frontend/src/api/stats.ts
import { apiClient } from "./client";

export interface ProcessStatsResponse {
  warn_bytes: number;
  danger_bytes: number;
  items: WorktreeProcessStats[];
}

export interface WorktreeProcessStats {
  repo: string;
  wt_name: string;
  total_rss_bytes: number;
  level: "ok" | "warn" | "danger";
  services: ServiceProcessStat[];
}

export interface ServiceProcessStat {
  name: string;
  pid: number;
  port: number;
  alive: boolean;
  procs: number;
  rss_bytes: number;
  uptime_sec: number;
}

export const statsApi = {
  list: (): Promise<ProcessStatsResponse> =>
    apiClient.get("api/process-stats").json(),
};

// 稼働時間の人間可読表示（例: 45s / 12m / 3h05m / 2d04h）。
export function formatDuration(sec: number): string {
  if (sec < 60) return `${sec}s`;
  if (sec < 3600) return `${Math.floor(sec / 60)}m`;
  if (sec < 86400) {
    const h = Math.floor(sec / 3600);
    const m = Math.floor((sec % 3600) / 60);
    return `${h}h${String(m).padStart(2, "0")}m`;
  }
  const d = Math.floor(sec / 86400);
  const h = Math.floor((sec % 86400) / 3600);
  return `${d}d${String(h).padStart(2, "0")}h`;
}

export function formatBytes(bytes: number): string {
  if (bytes === 0) return "0B";
  const k = 1024;
  const sizes = ["B", "K", "M", "G", "T", "P"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + sizes[i];
}

File: internal/handler/stats.go
package handler

import (
	"net/http"
	"os"

	"wt/internal/wt/devserver"
	"wt/internal/wt/procstats"
	"wt/internal/wt/settings"
	"wt/internal/wt/tree"
)

// procRoot は procstats が読む /proc のルート。実運用では常に "/proc"。
const procRoot = "/proc"

type ProcessStatsResponse struct {
	WarnBytes        uint64                 `json:"warn_bytes"`
	DangerBytes      uint64                 `json:"danger_bytes"`
	TotalRSSBytes    uint64                 `json:"total_rss_bytes"`
	InotifyInstances int                    `json:"inotify_instances"`
	InotifyMax       int                    `json:"inotify_max"`
	Items            []WorktreeProcessStats `json:"items"`
}

type WorktreeProcessStats struct {
	Repo          string               `json:"repo"`
	WtName        string               `json:"wt_name"`
	TotalRSSBytes uint64               `json:"total_rss_bytes"`
	Level         string               `json:"level"`
	Services      []ServiceProcessStat `json:"services"`
}

type ServiceProcessStat struct {
	Name      string `json:"name"`
	PID       int    `json:"pid"`
	Port      int    `json:"port"`
	Alive     bool   `json:"alive"`
	Procs     int    `json:"procs"`
	RSSBytes  uint64 `json:"rss_bytes"`
	UptimeSec int64  `json:"uptime_sec"`
}

func calcLevel(total, warn, danger uint64) string {
	if total >= danger {
		return "danger"
	}
	if total >= warn {
		return "warn"
	}
	return "ok"
}

func (h *Handler) GetProcessStats(w http.ResponseWriter, r *http.Request) {
	st := settings.Load()
	warnBytes := uint64(st.ProcessStats.WarnMB) * 1024 * 1024
	dangerBytes := uint64(st.ProcessStats.DangerMB) * 1024 * 1024

	resp := ProcessStatsResponse{
		WarnBytes:   warnBytes,
		DangerBytes: dangerBytes,
		Items:       []WorktreeProcessStats{}, // initialized so it renders as [] not null
	}

	entries := tree.Entries()

	var snapshot map[int]procstats.GroupStat
	snapshotFetched := false

	for i := range entries {
		entry := &entries[i]
		recorded := devserver.Recorded(entry.Path)
		if len(recorded) == 0 {
			continue
		}

		if !snapshotFetched {
			snapshot, _ = procstats.Snapshot(procRoot)
			snapshotFetched = true
		}

		wtStats := WorktreeProcessStats{
			Repo:     entry.Repo,
			WtName:   entry.WtName,
			Services: []ServiceProcessStat{}, // initialize to []
		}

		for _, rec := range recorded {
			svcStat := ServiceProcessStat{
				Name: rec.Name,
				PID:  rec.PID,
				Port: rec.Port,
			}
			if gs, ok := snapshot[rec.PID]; ok && gs.Procs > 0 {
				svcStat.Alive = true
				svcStat.Procs = gs.Procs
				svcStat.RSSBytes = gs.RSSBytes
				svcStat.UptimeSec = gs.UptimeSec
				wtStats.TotalRSSBytes += gs.RSSBytes
			} else {
				svcStat.Alive = false
				svcStat.Procs = 0
				svcStat.RSSBytes = 0
				svcStat.UptimeSec = 0
			}
			wtStats.Services = append(wtStats.Services, svcStat)
		}

		wtStats.Level = calcLevel(wtStats.TotalRSSBytes, warnBytes, dangerBytes)
		resp.Items = append(resp.Items, wtStats)
		resp.TotalRSSBytes += wtStats.TotalRSSBytes
	}

	resp.InotifyInstances, resp.InotifyMax = procstats.InotifyInstances(procRoot, os.Getuid())

	jsonOK(w, resp)
}
